### PR TITLE
OS interfaces

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following adds the `/servers/{server_id}/os-interface` endpoint
for listing all ports associated with a server.

Added subsequent tests to the stubs to ensure that we test the getting
the expected interfaces back from the stub server.

The is required to fulfill the need from Juju to be able to get all the
ports of a given server, to then be able to delete them when destroying
a controller.

#### Refrence

The documentation for this is here:
https://docs.openstack.org/api-ref/compute/?expanded=list-port-interfaces-detail

### CODE REVIEW

File: nova/live_test.go [hunk 1]
```diff
@@ -747,3 +747,30 @@ func (s *LiveTests) TestCreateServerBlockDeviceMappingVolume(c *gc.C) {
...
+		Name:             "inst-os-interface",
+		FlavorId:         s.testFlavorId,
+		ImageId:          s.testImageId,
+		AvailabilityZone: s.testAvailabilityZone,
+		Networks: []nova.ServerNetworks{{
+			NetworkId: s.testNetwork,
+		}},
+	})
+	c.Assert(err, gc.IsNil)
+	defer s.nova.DeleteServer(entity.Id)
```
Comment (hmlanigan) [inline nova/live_test.go]: ignored returned error

File: nova/nova.go [hunk 2]
```diff
@@ -590,3 +590,38 @@ func (c *Client) SetServerMetadata(serverId string, metadata map[string]string)
...
+
+// OSInterface represents an interface attachment to a server.
+type OSInterface struct {
+	FixedIPs   []PortFixedIP `json:"fixed_ips,omitempty"`
+	IPAddress  string        `json:"ip_address"`
+	MacAddress string        `json:"mac_addr,omitempty"`
+	NetID      string        `json:"net_id,omitempty"`
+	PortID     string        `json:"port_id,omitempty"`
+	PortState  string        `json:"port_state,omitempty"`
+}
```
Comment (hmlanigan) [inline nova/nova.go:610]: Just for completeness.

```suggestion
	Tag         string        `json:"tag,omitempty"`
}
```

File: nova/nova.go [hunk 3]
```diff
@@ -590,3 +590,38 @@ func (c *Client) SetServerMetadata(serverId string, metadata map[string]string)
...
+
+// ListOSInterfaces lists all the os-interfaces (port interfaces) associated
+// with a given server.
+// https://docs.openstack.org/api-ref/compute/?expanded=list-port-interfaces-detail
+func (c *Client) ListOSInterfaces(serverId string) ([]OSInterface, error) {
```
Comment (hmlanigan) [inline nova/nova.go:615]: Would it be useful to add call to encapsulate: Show Port Interface Details?  /servers/{server_id}/os-interface/{port_id}.  It provides slightly more data.

File: nova/nova.go [hunk 4]
```diff
@@ -590,3 +590,38 @@ func (c *Client) SetServerMetadata(serverId string, metadata map[string]string)
...
+	var resp struct {
+		InterfaceAttachments []OSInterface `json:"interfaceAttachments"`
+	}
+	requestData := goosehttp.RequestData{
+		RespValue: &resp,
+	}
```
Comment (hmlanigan) [inline nova/nova.go:622]:
```suggestion
            ExpectedStatus: []int{http.StatusOK},
	}
```

**Review by hmlanigan — APPROVED** [review]
A few suggestions, but overall LGTM

File: nova/nova.go (hunk 3 above)
Comment (SimonRichardson) [reply nova/nova.go:615]: I did think about doing that, I'll make a follow-up card to look at that in another PR.

**Pull request merged**